app/v2/controllers/V2MovementsController.scala

```diff
@@ -338,10 +339,21 @@ class V2MovementsControllerImpl @Inject() (
   def attachLargeMessage(movementId: MovementId, messageId: MessageId): Action[JsValue] =
     Action.async(parse.json) {
       implicit request =>
-        parseAndLogUpscanResponse(request.body).fold[Result](
-          presentationError => Status(presentationError.code.statusCode)(Json.toJson(presentationError)),
-          _ => Ok
-        )
+        implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequest(request)
+        parseAndLogUpscanResponse(request.body)
+          .map {
+            upscanResponse =>
+              (for {
+                objectStoreResponse <- objectStoreService.addMessage(upscanResponse.downloadUrl.get, movementId, messageId).asPresentation
+              } yield objectStoreResponse).fold[Result](
+                _ => Ok, //TODO: Send notification to PPNS with details of the error
+                _ => Ok  //TODO: Send notification to PPNS with details of the success
+              )
+          }
+          .fold[Result](
```

Comment:
A bit nitpicky, but it took me a moment to understand why you're `map`ping here rather than parsing the upscan response in the for comprehension -- I get it now in that if Upscan sends rubbish, we need to tell them as such.

However, if we have a failed `UpscanResponse`, this code is going to fall over because of `upscanResponse.downloadUrl.get`. That needs to be handled.

Reply:
In parseAndLogUpscanResponse we ensure that we have the download URL for a successful callback, otherwise we send them 400 - that's something I spotted in Prasanth's PR and it needs to also be updated in the router.

Reply:
My bad, you're right!